src/h3/mod.rs

```diff
@@ -26,4 +26,778 @@
+            is_server,
+
+            highest_request_stream_id: 0,
+            highest_uni_stream_id: initial_uni_stream_id,
+
+            streams: StreamMap::new(),
+
+            local_settings: ConnectionSettings {
+                num_placeholders: Some(config.num_placeholders),
+                max_header_list_size: Some(config.max_header_list_size),
+                qpack_max_table_capacity: Some(config.qpack_max_table_capacity),
+                qpack_blocked_streams: Some(config.qpack_blocked_streams),
+            },
+
+            peer_settings: ConnectionSettings {
+                num_placeholders: None,
+                max_header_list_size: None,
+                qpack_max_table_capacity: None,
+                qpack_blocked_streams: None,
+            },
+
+            control_stream_id: None,
+            peer_control_stream_id: None,
+
+            qpack_encoder: qpack::Encoder::new(),
+            qpack_decoder: qpack::Decoder::new(),
+
+            local_qpack_streams: QpackStreams {
+                encoder_stream_id: None,
+                decoder_stream_id: None,
+            },
+            peer_qpack_streams: QpackStreams {
+                encoder_stream_id: None,
+                decoder_stream_id: None,
+            },
+        })
+    }
+
+    /// Test if all local and peer Control and QPACK streams are open
+    pub fn is_established(&self) -> bool {
+        self.control_stream_id.is_some() &&
+            self.local_qpack_streams.encoder_stream_id.is_some() &&
+            self.local_qpack_streams.decoder_stream_id.is_some() &&
+            self.peer_control_stream_id.is_some() &&
+            self.peer_qpack_streams.decoder_stream_id.is_some() &&
+            self.peer_qpack_streams.encoder_stream_id.is_some()
+    }
+
+    /// Prepare a request in HTTP/3 wire format, allocate a stream ID and send
+    /// it.
+    pub fn send_request_headers(
+        &mut self, quic_conn: &mut super::Connection, headers: &[qpack::Header],
+        has_body: bool,
+    ) -> Result<(u64)> {
+        let mut d = [42; 65535];
+
+        let headers_len = headers
+            .iter()
+            .fold(0, |acc, h| acc + h.value().len() + h.name().len() + 32);
+        let mut header_block = vec![0; headers_len];
+        let len = self.qpack_encoder.encode(&headers, &mut header_block)?;
+        header_block.truncate(len);
+
+        let req_frame = frame::Frame::Headers { header_block };
+
+        let mut b = octets::Octets::with_slice(&mut d);
+        req_frame.to_bytes(&mut b)?;
+
+        let stream_id = self.get_available_request_stream()?;
+        self.streams
+            .insert(stream_id, stream::Stream::new(stream_id, !has_body));
+
+        let off = b.off();
+
+        trace!(
+            "{} sending request of size {} on stream {}",
+            quic_conn.trace_id(),
+            off,
+            stream_id
+        );
+
+        if let Err(e) = quic_conn.stream_send(stream_id, &d[..off], true) {
```

Comment:
```suggestion
        quic_conn.stream_send(stream_id, &d[..off], fin)?;
```

and then remove the next 3 lines as well.